library/std/src/sys/pal/windows/api.rs

```diff
@@ -291,3 +292,49 @@ impl WinError {
     pub const TIMEOUT: Self = Self::new(c::ERROR_TIMEOUT);
     // tidy-alphabetical-end
 }
+
+/// A wrapper around a UNICODE_STRING that is equivalent to `&[u16]`.
+///
+/// It is preferable to use `from_slice_with_nul` for our own strings as a
+/// mitigation for #143078.
+#[derive(Copy, Clone)]
+pub struct UnicodeStrRef<'a> {
```

Comment:
Could you mention in the docs something about this optionally being nul-terminated but that not being a guarantee? To help explain why the functions aren't `unsafe`.